src/rust/engine/dep_inference/src/python/tests.rs

```diff
@@ -407,6 +407,119 @@ fn weak_imports() {
     &["one.two.three"],
     &[],
   );
+  // Ensure we preserve the stack of weakens with try-except
+  assert_imports_strong_weak(
+    r"
+    try:
+        with suppress(ImportError):
+            import weak0
+        import weak1
+    except ImportError:
+        with suppress(ImportError):
+            import weak2
+        import strong0
+    import strong1
+    ",
+    &["strong0", "strong1"],
+    &["weak0", "weak1", "weak2"],
+  );
+}
+
+#[test]
+fn contextlib_suppress_weak_imports() {
+  // standard contextlib.suppress
+  assert_imports_strong_weak(
+    r"
+    with contextlib.suppress(ImportError):
+        import weak0
+    ",
+    &[],
+    &["weak0"],
+  );
+  // ensure we reset the weakened status
+  assert_imports_strong_weak(
+    r"
+    with contextlib.suppress(ImportError):
+        import weak0
+
+    import strong0
+    ",
+    &["strong0"],
+    &["weak0"],
+  );
+  // Allow other error types to be suppressed
+  assert_imports_strong_weak(
+    r"
+    with suppress(NameError, ImportError):
+        import weak0
+    ",
+    &[],
+    &["weak0"],
+  );
+  // We should respect the intention of any function that is obviously suppressing ImportErrors
+  assert_imports_strong_weak(
+    r"
+    with suppress(ImportError):
+        import weak0
+    ",
+    &[],
+    &["weak0"],
+  );
+  // We should not weaken because of other suppressions
+  assert_imports_strong_weak(
+    r"
+    with contextlib.suppress(NameError):
+        import strong0
+      ",
+    &["strong0"],
+    &[],
+  );
+  // Ensure we preserve the stack of weakens
+  assert_imports_strong_weak(
+    r"
+    with suppress(ImportError):
+        import weak0
+        with suppress(ImportError):
+            import weak1
+        import weak2
+    ",
+    &[],
+    &["weak0", "weak1", "weak2"],
+  );
+  // Ensure we preserve the stack of weakens with try-except
+  assert_imports_strong_weak(
+    r"
+    with suppress(ImportError):
+        try:
+            import weak0
+        except ImportError:
+            import weak1
+        import weak2
+    ",
+    &[],
+    &["weak0", "weak1", "weak2"],
+  );
+  // Ensure we aren't affected by weirdness in tree-sitter
+  // where in the viewer the second import wasn't assigned the correct parent
+  assert_imports_strong_weak(
+    r"
+    with suppress(ImportError):
+        import weak0
+        import weak1
+    ",
+    &[],
+    &["weak0", "weak1"],
+  );
+  // Ensure that we still traverse withitems
+  let withitems_open = r"
+    with (
+        open('/dev/null') as f0,
+        open('data/subdir1/a.json') as f1,
+    ):
+        pass
+    ";
+  assert_imports_strong_weak(withitems_open, &[], &[]);
+  assert_strings(withitems_open, &["/dev/null", "data/subdir1/a.json"])
 }
```

Comment:
Same comment as up there, maybe a few more tests.

Additionally, we probably want a test for `with suppress(ImportError) as e:` for shits and giggles.

And please do throw any other wacky thing you can think of. The tree-sitter code is easy to get right on well-formed constructs. Then we go kaboom when we weren't creative.

So stuff like:
```python
with suppress(ImportError) as e:
    ....

with open() as f, suppress(ImportError):
    ....

# I think this is supported as of Python 3.11 or something
with (suppress(ImportError),):
    ....
```

idk get creative :stuck_out_tongue:

Reply:
man some of those are cursed. The grammar itself has an unreleased fix for the 3rd one. I'm not sure if I should work around that or just wait for a new minor release of tree-sitter-python.